jetstream/analysis.py

```diff
@@ -384,24 +384,66 @@ def subset_metric_table(
         self,
         metrics_table_name: str,
         segment: str,
-        metric: Metric,
+        summary: metric.Summary,
         analysis_basis: AnalysisBasis,
+        period: AnalysisPeriod,
     ) -> DataFrame:
         """Pulls the metric data for this segment/analysis basis"""
 
         query = self._create_subset_metric_table_query(
-            metrics_table_name, segment, metric, analysis_basis
+            metrics_table_name, segment, summary, analysis_basis, period
         )
 
         results = self.bigquery.execute(query).to_dataframe()
 
         return results
 
-    @staticmethod
     def _create_subset_metric_table_query(
-        metrics_table_name: str, segment: str, metric: Metric, analysis_basis: AnalysisBasis
+        self,
+        metrics_table_name: str,
+        segment: str,
+        summary: metric.Summary,
+        analysis_basis: AnalysisBasis,
+        period: AnalysisPeriod,
+    ) -> str:
+        if covariate_params := summary.statistic.params.get("covariate_adjustment", False):
+            covariate_metric_name = covariate_params["metric"]
+            covariate_period = AnalysisPeriod(covariate_params["period"])
+            if covariate_period != period:
+                query = self._create_subset_metric_table_query_covariate(
+                    metrics_table_name,
+                    segment,
+                    summary.metric,
+                    analysis_basis,
+                    covariate_period,
+                    covariate_metric_name,
+                )
+            else:
+                # when we configure a metric, all statistics are applied to all periods
+                # however, to perform covariate adjustment we must use data from a different
+                # period. So the metric will be configured with analysis periods like
+                # [preenrollment_week, weekly, overall] but covariate adjustment should
+                # only be applied on weekly and overall when using preenrollment_week
+                # as the covariate.
+                query = self._create_subset_metric_table_query_univariate(
+                    metrics_table_name, segment, summary.metric, analysis_basis
+                )
+        else:
+            query = self._create_subset_metric_table_query_univariate(
+                metrics_table_name, segment, summary.metric, analysis_basis
+            )
+
+        return query
```

Comment:
Do we need to initialize the `query` variable first? Otherwise it would throw: `UnboundLocalError: local variable 'query' referenced before assignment`.

Reply:
Made this change